File: app/src/main/java/com/fairshare/domain/model/sync/OpPayload.kt
package com.fairshare.domain.model.sync

import kotlinx.serialization.Serializable

/**
 * Payload of an [Operation]. Each variant fully describes the mutation it
 * applies to the materialized state.
 *
 * For Upsert variants the snapshot is authoritative: the materializer keeps the
 * variant with the highest `(lamport, deviceId)` tuple per `entityId` and
 * writes its snapshot to the local DB (see DESIGN.md §4.2).
 *
 * Delete variants act as tombstones suppressing earlier Upserts for the same
 * entity, until a later Upsert revives it.
 */
@Serializable
sealed interface OpPayload {

    @Serializable
    data class EventUpsert(val event: EventSnapshot) : OpPayload

    @Serializable
    data class EventDelete(val eventId: String) : OpPayload

    @Serializable
    data class ParticipantUpsert(val participant: ParticipantSnapshot) : OpPayload

    @Serializable
    data class ParticipantDelete(val participantId: String) : OpPayload

    @Serializable
    data class ExpenseUpsert(val expense: ExpenseSnapshot) : OpPayload

    @Serializable
    data class ExpenseDelete(val expenseId: String) : OpPayload

    /**
     * Upsert / delete for a user-created custom Category. Default
     * categories are hardcoded ([com.fairshare.domain.model.DefaultCategories])
     * and never travel through this log.
     */
    @Serializable
    data class CategoryUpsert(val category: CategorySnapshot) : OpPayload

    @Serializable
    data class CategoryDelete(val categoryId: String) : OpPayload
}

/**
 * Returns the id of the entity this payload mutates. Used by the materializer
 * to group ops by entity for LWW resolution.
 */
val OpPayload.entityId: String
    get() = when (this) {
        is OpPayload.EventUpsert -> event.id
        is OpPayload.EventDelete -> eventId
        is OpPayload.ParticipantUpsert -> participant.id
        is OpPayload.ParticipantDelete -> participantId
        is OpPayload.ExpenseUpsert -> expense.id
        is OpPayload.ExpenseDelete -> expenseId
        is OpPayload.CategoryUpsert -> category.id
        is OpPayload.CategoryDelete -> categoryId
    }

/**
 * Entity family this payload belongs to. LWW resolution is performed
 * independently per family per entityId.
 */
enum class EntityKind { EVENT, PARTICIPANT, CATEGORY, EXPENSE }

val OpPayload.entityKind: EntityKind
    get() = when (this) {
        is OpPayload.EventUpsert,
        is OpPayload.EventDelete -> EntityKind.EVENT
        is OpPayload.ParticipantUpsert,
        is OpPayload.ParticipantDelete -> EntityKind.PARTICIPANT
        is OpPayload.CategoryUpsert,
        is OpPayload.CategoryDelete -> EntityKind.CATEGORY
        is OpPayload.ExpenseUpsert,
        is OpPayload.ExpenseDelete -> EntityKind.EXPENSE
    }

File: webapp/src/core/sync/opPayload.ts
import type {
  CategorySnapshot,
  EventSnapshot,
  ExpenseSnapshot,
  ParticipantSnapshot,
} from "./snapshots";

/**
 * `OpPayload` from the Kotlin side. The `type` discriminator value is
 * the *fully-qualified Kotlin class name* of the variant — that's how
 * `kotlinx.serialization`'s `classDiscriminator` defaults work for
 * sealed types. These constants must not drift from Kotlin or Android
 * peers will fail to decode our ops.
 */

export const OP_TYPE = {
  EventUpsert: "com.fairshare.domain.model.sync.OpPayload.EventUpsert",
  EventDelete: "com.fairshare.domain.model.sync.OpPayload.EventDelete",
  ParticipantUpsert:
    "com.fairshare.domain.model.sync.OpPayload.ParticipantUpsert",
  ParticipantDelete:
    "com.fairshare.domain.model.sync.OpPayload.ParticipantDelete",
  ExpenseUpsert: "com.fairshare.domain.model.sync.OpPayload.ExpenseUpsert",
  ExpenseDelete: "com.fairshare.domain.model.sync.OpPayload.ExpenseDelete",
  CategoryUpsert: "com.fairshare.domain.model.sync.OpPayload.CategoryUpsert",
  CategoryDelete: "com.fairshare.domain.model.sync.OpPayload.CategoryDelete",
} as const;

export type OpType = (typeof OP_TYPE)[keyof typeof OP_TYPE];

export type OpPayload =
  | { type: typeof OP_TYPE.EventUpsert; event: EventSnapshot }
  | { type: typeof OP_TYPE.EventDelete; eventId: string }
  | { type: typeof OP_TYPE.ParticipantUpsert; participant: ParticipantSnapshot }
  | { type: typeof OP_TYPE.ParticipantDelete; participantId: string }
  | { type: typeof OP_TYPE.ExpenseUpsert; expense: ExpenseSnapshot }
  | { type: typeof OP_TYPE.ExpenseDelete; expenseId: string }
  | { type: typeof OP_TYPE.CategoryUpsert; category: CategorySnapshot }
  | { type: typeof OP_TYPE.CategoryDelete; categoryId: string };

export type EntityKind = "EVENT" | "PARTICIPANT" | "CATEGORY" | "EXPENSE";

export function entityIdOf(p: OpPayload): string {
  switch (p.type) {
    case OP_TYPE.EventUpsert:
      return p.event.id;
    case OP_TYPE.EventDelete:
      return p.eventId;
    case OP_TYPE.ParticipantUpsert:
      return p.participant.id;
    case OP_TYPE.ParticipantDelete:
      return p.participantId;
    case OP_TYPE.ExpenseUpsert:
      return p.expense.id;
    case OP_TYPE.ExpenseDelete:
      return p.expenseId;
    case OP_TYPE.CategoryUpsert:
      return p.category.id;
    case OP_TYPE.CategoryDelete:
      return p.categoryId;
  }
}

export function entityKindOf(p: OpPayload): EntityKind {
  switch (p.type) {
    case OP_TYPE.EventUpsert:
    case OP_TYPE.EventDelete:
      return "EVENT";
    case OP_TYPE.ParticipantUpsert:
    case OP_TYPE.ParticipantDelete:
      return "PARTICIPANT";
    case OP_TYPE.CategoryUpsert:
    case OP_TYPE.CategoryDelete:
      return "CATEGORY";
    case OP_TYPE.ExpenseUpsert:
    case OP_TYPE.ExpenseDelete:
      return "EXPENSE";
  }
}
